lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,443 @@
+	}
+
+	// We use two transactions here to preserve the property that our
+	// leveldb wrapper has, that writes in a transaction are completely
+	// invisible until it's committed, even inside that same transaction.
+	rtxn := b.bdb.NewTransaction(false)
+	wtxn := b.bdb.NewTransaction(true)
+	return &badgerTransaction{
+		badgerSnapshot: badgerSnapshot{
+			txn: rtxn,
+			rel: rel1,
+		},
+		txn: wtxn,
+		bdb: b.bdb,
+		rel: rel2,
+	}, nil
+}
+
+func (b *badgerBackend) Close() error {
+	b.closeWG.CloseWait()
+	return wrapBadgerErr(b.bdb.Close())
+}
+
+func (b *badgerBackend) Get(key []byte) ([]byte, error) {
+	if err := b.closeWG.Add(1); err != nil {
+		return nil, err
+	}
+	defer b.closeWG.Done()
+
+	txn := b.bdb.NewTransaction(false)
+	defer txn.Discard()
+	item, err := txn.Get(key)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	val, err := item.ValueCopy(nil)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	return val, nil
+}
+
+func (b *badgerBackend) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	if err := b.closeWG.Add(1); err != nil {
+		return nil, err
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerPrefixIterator(txn, prefix)
+	it.releaseFn = func() {
+		defer b.closeWG.Done()
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) NewRangeIterator(first, last []byte) (Iterator, error) {
+	if err := b.closeWG.Add(1); err != nil {
+		return nil, err
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerRangeIterator(txn, first, last)
+	it.releaseFn = func() {
+		defer b.closeWG.Done()
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) Put(key, val []byte) error {
+	if err := b.closeWG.Add(1); err != nil {
+		return err
+	}
+	defer b.closeWG.Done()
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Set(key, val); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Delete(key []byte) error {
+	if err := b.closeWG.Add(1); err != nil {
+		return err
+	}
+	defer b.closeWG.Done()
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Delete(key); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Compact() error {
+	if err := b.closeWG.Add(1); err != nil {
+		return err
+	}
+	defer b.closeWG.Done()
+
+	// This weird looking loop is as recommended in the README
+	// (https://github.com/dgraph-io/badger/tree/v2.0.3#garbage-collection).
+	// Basically, the RunValueLogGC will pick some promising thing to
+	// garbage collect at random and return nil if it improved the
+	// situation, then return ErrNoRewrite when there is nothing more to GC.
+	// The 0.5 is the discard ratio, for which the method docs say they
+	// "recommend setting discardRatio to 0.5, thus indicating that a file
+	// be rewritten if half the space can be discarded".
+	var err error
+	t0 := time.Now()
+	for err == nil {
+		if time.Since(t0) > time.Hour {
```

Comment:
Hour?
That's a long time.

I guess we rarely call this so it should not matter?

Also, this would never be printed if it just hangs.

Reply:
Yeah... I mean, it can probably hang if the disk hangs, but that goes for all the other database calls as well and is not something we generally worry about. This is more for the really unexpected case of badger changing their api and this looping forever for whatever unforeseen reason. (Or compaction just honestly not able to keep up with our writes for a long time.) Compaction currently just runs after schema updates and gc runs.